synedrion/src/uint/signed.rs

```diff
@@ -332,76 +435,447 @@ where
+            let signed = Signed::new_positive(v, self.bound);
+            if let Some(signed) = signed {
+                CtOption::new(signed, 1u8.into())
+            } else {
+                CtOption::new(Signed::default(), 0u8.into())
+            }
+        })
+    }
+}
+
+impl<T> Sub<Signed<T>> for Signed<T>
+where
+    T: crypto_bigint::Bounded + ConditionallySelectable + Encoding + Integer,
+{
     type Output = Self;
     fn sub(self, rhs: Self) -> Self::Output {
-        self.checked_add(&-rhs).unwrap()
+        self.checked_add(&-rhs).expect("Invalid subtraction")
     }
 }
 
-impl<T: UintLike> Sub<&Signed<T>> for Signed<T> {
+impl<T> Sub<&Signed<T>> for Signed<T>
+where
+    T: crypto_bigint::Bounded + ConditionallySelectable + Encoding + Integer,
+{
     type Output = Self;
     fn sub(self, rhs: &Self) -> Self::Output {
-        self.checked_add(&-rhs).unwrap()
+        self.checked_add(&-rhs).expect("Invalid subtraction")
     }
 }
 
-impl<T: UintLike> Mul<Signed<T>> for Signed<T> {
+impl<T> Mul<Signed<T>> for Signed<T>
+where
+    T: Integer + Encoding + crypto_bigint::Bounded + ConditionallySelectable,
+{
     type Output = Self;
     fn mul(self, rhs: Self) -> Self::Output {
         self.checked_mul(&rhs).unwrap()
     }
 }
 
-impl<T: UintLike> Mul<&Signed<T>> for Signed<T> {
+impl<T> Mul<&Signed<T>> for Signed<T>
+where
+    T: Integer + Encoding + crypto_bigint::Bounded + ConditionallySelectable,
+{
     type Output = Self;
     fn mul(self, rhs: &Self) -> Self::Output {
         self.checked_mul(rhs).unwrap()
     }
 }
 
-impl<T: UintLike> core::iter::Sum for Signed<T> {
+impl<T> core::iter::Sum for Signed<T>
+where
+    T: Integer + Encoding + crypto_bigint::Bounded,
+{
     fn sum<I: Iterator<Item = Self>>(iter: I) -> Self {
         iter.reduce(|x, y| x.checked_add(&y).unwrap())
             .unwrap_or(Self::default())
     }
 }
 
-impl<'a, T: UintLike> core::iter::Sum<&'a Self> for Signed<T> {
+impl<'a, T> core::iter::Sum<&'a Self> for Signed<T>
+where
+    T: Integer + Encoding + crypto_bigint::Bounded,
+{
     fn sum<I: Iterator<Item = &'a Self>>(iter: I) -> Self {
         iter.cloned().sum()
     }
 }
+
+// @reviewers: Implementing `PartialEq`/`Eq` this way disregards the bound when it comes to equality,
+// which is a major difference to the implementations that were `#[derive]`d before. I think it is
+// more intuitive and useful to be able to compare two `Signed` with different bounds but I can
+// totally see how others may think differently, so this is definitely something we should discuss.
+impl<T> PartialEq for Signed<T>
+where
+    T: crypto_bigint::Bounded + ConditionallySelectable + Encoding + Integer,
+{
+    fn eq(&self, other: &Self) -> bool {
+        self.abs() == other.abs() && self.is_negative().ct_eq(&other.is_negative()).into()
+    }
+}
+impl<T> Eq for Signed<T> where
+    T: crypto_bigint::Bounded + ConditionallySelectable + Encoding + Integer
+{
+}
+
+impl<T> PartialOrd for Signed<T>
```

Comment:
Do we need `Signed` to be `Ord` anywhere?

Reply:
We do not, this is indeed not strictly necessary. 

I added it because it made writing tests a bit easier and because it made using `Signed` values more like a "normal" number type.

Do you think I should remove it?

Reply:
If it's just for tests, perhaps decorate it with `#[cfg(test)]`